### ISSUE TITLE
iOS 12.2 not working

### ISSUE THREAD

**Issue by danigilme** [richtr/NoSleep.js#75]
* Device Name: iPhone X
* Device Version: 12.2
* Browser Name: Safari
* Browser Version: ?
* NoSleep.js Version: Last one.

### Expected Behavior
Prevent to sleep my device with the official DEMO.

### Actual Behavior
I tried the demo on an iPhone X with Safari on iOS 12.2 and is not working at all. I check the player and is not playing any video or smth.

### Code
Just the official demo. I’ve not tried any extra code.

### How Do We Reproduce?
With an iPhone with iOS 12.2 and Safari, checking the github DEMO.

**Comment by beshur** [comment]
Me too, on iPad 12.2

**Comment by ngerritsen** [comment]
Personally I only have it not working when I add the site to my homescreen as a fullscreen app and re-open it.

**Comment by ngerritsen** [comment]
Just copying what I wrote on the other bug on iOS, just my findings:

> > I am having the same issue. It ONLY works right after I added the app to the homescreen, but it doesn’t if I close and open it again.
> 
> Edit: It’s getting more interesting, I actually have that same problem if I do not add it to my homescreen, if I re-open iOS safari (without doing a hard reload) it doesn’t work anymore... I have the idea the dummy video is “stuck” or not able to play anymore.
> 
> Edit2: did some more investigating, it seems it’s actually working always if I reconstruct nosleep before enabling it every time, so like
> 
> ```js
> let nosleep = null;
> 
> function enable() {
>   if (nosleep) nosleep.disable(); // Just to be sure if you forgot to disable.
>   nosleep = new NoSleep();
>   nosleep.enable();
> }
> 
> function disable() {
>   nosleep.disable();
> }
> ```
> 
> It seems to fix the problems I had, but I’m just giving the information I now have, I’m not sure if it’s the actual solution.

**Comment by beshur** [comment]
@ngerritsen, it may happen due to `let` block scoping, if you declare `nosleep` in some other block, far from `enable` function body. Or you posted complete of your code?

**Comment by ngerritsen** [comment]
@beshur what would happen due to let scoping? The example I posted actually fixes the issue, because I create a new instance every time. Nothing to do with scoping.

**Comment by beshur** [comment]
@ngerritsen, I was saying if in your previous realisation `nosleep` instance was declared not within the `enable`/`disable` scope, that may be an issue.

Anyway, even the demo page does not work for me.

**Comment by ngerritsen** [comment]
@beshur yes because the example probably doesn’t recreate the instance, because this is not how the documentation states you should do it. Maybe try my workaround, can you show your code? Are you sure you enable synchronously inside you event listener?

**Comment by danigilme** [comment]
As far as I’ve seen, the demo is not working properly. Because I know a page which uses it (version 0.7.0) and it’s working perfectly. So I think the problem is with the demo page 😅.

**Comment by ngerritsen** [comment]
@danigilme the problem is not with the demo page, the problem is that when you do not create a new instance of `NoSleep` after you have navigated back to a iOS Safari tab, enable doesn’t work anymore. Only after you created a new instance. Then doing `(new NoSleep()).enable()` works in that case, probably because it recreates the video.

My preliminary conclusion: _you need to recreate the html5 video to make it consistently work._

**Comment by danigilme** [comment]
Okay thanks then. When I can I will try it and hope it works 🙈

**Comment by beshur** [comment]
Guys, I'm not sure what happened, but both the demo page and my code work now.
Did not use @ngerritsen suggestion of creating new instances on NoSleep every time.

However I slightly modified the library itself - added an `isEnabled` method, which returns `! this.noSleepVideo.paused`, and I only call `enable` if it's not yet enabled.

**Comment by danigilme** [comment]
Great to know that! 🙌

**Comment by ngerritsen** [comment]
@beshur sounds good, would be nice if that could be the solution.

**Comment by Domratchev** [comment]
I believe this PR fixes this issue:
https://github.com/ViRPo/NoSleep.js/pull/1

**Comment by mhoeneveld** [comment]
I solved it by using the document.visibilityState;
Indeed on iOS the noSleep function is disabled as soon as the PWA goes to the background. The rest of the site works fine after moving it to the forground again but not the noSleep function.
By detecting the document.visibilityState you can or reload the page or just enable wakeLockEnabled.

```
	<script>
		const state = document.visibilityState;
		
		document.addEventListener('visibilitychange', () => {
		  console.log(document.visibilityState);
		  if ( document.visibilityState == "visible" ) {
		  	//reload if the page comes back to visible
		  	window.location.reload();   //Reload the whole page or use the commands below.
                       //          noSleep.enable(); // keep the screen on!
                       //          wakeLockEnabled = true;		  	
		  }
		})
	</script>
```

**Comment by bobvanvliet** [comment]
The fix proposed by @Domratchev worked for me (iOS 12.3.1 on iPad mini 2).

**Comment by thenameisdavid** [comment]
To all of you who got this finally working: are you running the fix on NoSleep client side or server side? Strictly in browser or any webview users?

**Comment by bobvanvliet** [comment]
> To all of you who got this finally working: are you running the fix on NoSleep client side or server side? Strictly in browser or any webview users?

Isn't this thing client-side by definition? Anyway, using it in a simple one-page html/css/js webapp.

**Comment by thenameisdavid** [comment]
Fair enough. I was hoping I could use this to bypass iOS 13 (background mode) webview restrictions which terminate <audio> embeds once the device goes into sleep/background/lock mode. Just to be clear, I'm specifically referring to WKWebView instances. Cheers!

**Comment by farQtech** [comment]
I tried the workaround posted by @ngerritsen and it didn't work for me on iPhone 7+ iOS 13. works well on Android/desktops (win/mac).

**Comment by richtr** [comment]
There are conflicting reports and solutions on this issue and it's quite hard to follow the recommendations.

> However I slightly modified the library itself - added an isEnabled method, which returns ! this.noSleepVideo.paused, and I only call enable if it's not yet enabled.

Following this reported fix, an `isEnabled` function has been added in 0f153f2f45c692ce3d2db1998cd9cec6bbd7ce45.

If you have additional issues then please open a new issue.

**Comment by KHromak** [comment]
Hi guys, video plays, but iphone still goes to sleep after 2 mins. (30 sec sleep in settings). Any thoughts? thx
ios 15.4.1

**Comment by copypasteho** [comment]
> Hi guys, video plays, but iphone still goes to sleep after 2 mins. (30 sec sleep in settings). Any thoughts? thx
> ios 15.4.1

Check actual background settings on the device for Safari or the app. If using XCode :  make sure you have "background audio play" enabled under settings.  Test the demo web app using Safari on your device and see if it actually works. If it doesn't work -- abandon all hope. This library has always been buggy.

**Comment by KHromak** [comment]
> 
issue was in too short length of the video (1.079 seconds)
some times if application laggy, video goes on end and stop. than no sleep turn off.

i make rewind timer shorter:
thats helps:

 // mp4 source
                    this.noSleepVideo.addEventListener('timeupdate', () => {
                        if (this.noSleepVideo.currentTime > 0.4) {
                            this.noSleepVideo.currentTime = Math.random() * 0.9
                        }
                    })
